Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,239 @@
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic
+
+namespace Week2
+
+/-
+  # Logic: Implication, universal quantifier, existential quantifier, negation
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+/-!
+  # Implication and universal quantifier
+-/
+
+-- To write `∀` type
+
+-- Lean treats the universal quantifier as an implication.
+-- Implications are grouped on the RIGHT (N.B. this is different from the usual convention)
+
+theorem my_lemma : ∀ x y ε : ℝ, 0 < ε → ε ≤ 1 → |x| < ε → |y| < ε → |x * y| < ε :=
+  sorry
+
+-- Lean is all about types, and checking that something is of the right type
+section
+-- Naively, you can think of types as sets. Here these variables are real numbers
+variable (a b δ : ℝ)
+-- Being of the type of a Proposition means being a proof
+variable (h₀ : 0 < δ) (h₁ : δ ≤ 1)
+variable (ha : |a| < δ) (hb : |b| < δ)
+
+-- An implication is something like a function from sets of proofs to set of proofs
+#check my_lemma a b δ
+#check my_lemma a b δ h₀ h₁
+#check my_lemma a b δ h₀ h₁ ha hb
+
+end
+
+
+-- To prove an implication statement, we use intro
+
+example: ∀ x y z : ℝ, x ≤ y → y ≤ z → x ≤ z := by
+  intro x y z h₀ h₁
+  -- Now we can proceed as usual
+  apply le_trans
+  · apply h₀
+  · apply h₁
+
+
+-- Using the universal quantifier, we define an upper bounded function
+def FnUb (f : ℝ → ℝ) (a : ℝ) : Prop :=
```

Comment:
I think this is the first time we've encountered `Prop`. I think we should say something like: '`Prop` is a type, just like `ℝ`, but its terms are logical statements that can be true or false, e.g. `1 < 2`, `∀ x : ℕ, x ≥ 0` or `2 + 2 = 5`. As we saw before, Lean treats these statements as types themselves, whose terms are proofs of the statement. Hence, if `h : 1 < 2`, then `h` is a proof that `1 < 2`. The type `2 + 2 = 5` has no terms.'

Reply:
I was planning on saying this during the demo, but it doesn't hurt having it written down